spring-data-relational/src/main/java/org/springframework/data/relational/repository/query/RelationalExampleMapper.java

```diff
@@ -64,92 +72,194 @@ public <T> Query getMappedExample(Example<T> example) {
 	 * {@link Query}.
 	 *
 	 * @param example
-	 * @param entity
+	 * @param persistentEntity
 	 * @return query
 	 */
-	private <T> Query getMappedExample(Example<T> example, RelationalPersistentEntity<?> entity) {
+	private <T> Query getMappedExample(Example<T> example, RelationalPersistentEntity<?> persistentEntity) {
 
 		Assert.notNull(example, "Example must not be null");
-		Assert.notNull(entity, "RelationalPersistentEntity must not be null");
+		Assert.notNull(persistentEntity, "RelationalPersistentEntity must not be null");
 
-		PersistentPropertyAccessor<T> propertyAccessor = entity.getPropertyAccessor(example.getProbe());
+		PersistentPropertyAccessor<T> probePropertyAccessor = persistentEntity.getPropertyAccessor(example.getProbe());
 		ExampleMatcherAccessor matcherAccessor = new ExampleMatcherAccessor(example.getMatcher());
 
-		final List<Criteria> criteriaBasedOnProperties = new ArrayList<>();
+		final List<Criteria> criteriaBasedOnProperties = buildCriteriaRecursive( //
+				persistentEntity, //
+				matcherAccessor, //
+				probePropertyAccessor //
+		);
 
-		entity.doWithProperties((PropertyHandler<RelationalPersistentProperty>) property -> {
+		// Criteria, assemble!
+		Criteria criteria = Criteria.empty();
 
-			if (property.isCollectionLike() || property.isMap()) {
-				return;
-			}
+		for (Criteria propertyCriteria : criteriaBasedOnProperties) {
 
-			if (matcherAccessor.isIgnoredPath(property.getName())) {
-				return;
+			if (example.getMatcher().isAllMatching()) {
+				criteria = criteria.and(propertyCriteria);
+			} else {
+				criteria = criteria.or(propertyCriteria);
 			}
+		}
+
+		return Query.query(criteria);
+	}
+
+	private <T> @NotNull List<Criteria> buildCriteriaRecursive( //
```

Comment:
Care to remove all those lingering `@NotNull` annotations?

Reply:
Sure, done